### PR TITLE
[CI] Optimize unit tests including adding stochasticity

### PR DESCRIPTION
**Pull request by hughperkins**
Issue: #

### Brief Summary

copilot:summary

### Walkthrough

copilot:walkthrough

### CODE REVIEW

File: tests/python/test_linalg.py [hunk 1]
```diff
@@ -434,7 +450,7 @@ def run():
     np.testing.assert_allclose(M @ inv_np, np.eye(n_), rtol=tol, atol=tol)
 
 
-@pytest.mark.parametrize("n", [5, 6, 7, 8, 9, 10, 11, 12])
+@pytest.mark.parametrize("n", [5, pytest.param(12, marks=pytest.mark.slow)])
```
[the remark this answers is not in the document]
Comment (hughperkins) [reply tests/python/test_linalg.py:464]: no, I dont think we have to test for all these intermediate sizes.

File: docs/source/user_guide/contributing.md [hunk 2]
```diff
@@ -2,25 +2,13 @@
 
 ## Good practice reminder
 
-* *testing*: Any new features or modified code should be tested. You have to run the test suite using `python tests/run_tests.py` which sets up the right test environment for `pytest`. CLI arguments are forwarded to `pytest`. Do not use `pytest` directly as it behaves differently. To see a per-file timing breakdown (useful for identifying slow test files), set `QD_FILE_TIMING=1` — e.g. `QD_FILE_TIMING=1 python tests/run_tests.py`. This is enabled by default in the Mac CI job and the results appear in the GitHub Actions job summary.
+* *testing*: Any new features or modified code should be tested. Run the test suite with `python tests/run_tests.py` — see [unit_testing.md](unit_testing.md) for how to scope by file / keyword / arch, how the `slow` and `sample` markers interact, and how to reproduce a single CI failure.
```
Comment (hughperkins) [inline docs/source/user_guide/contributing.md]: replace '*testing*: Any new features or modified code should be tested. Run the test suite with `python tests/run_tests.py` — see [unit_testing.md](unit_testing.md) for how to scope by file / keyword / arch, how the `slow` and `sample` markers interact, and how to reproduce a single CI failure.' with '*testing*: Any new features or modified code should be tested. see [unit_testing.md](unit_testing.md)'

File: docs/source/user_guide/unit_testing.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
+# Unit testing
+
+This page documents how to run, write, and tune the Quadrants Python unit test suite. For one-shot setup of the build / dev environment, see [contributing.md](contributing.md).
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: - "For one-shot setup of the build / dev environment, see [contributing.md](contributing.md)." => "For setup of the build / dev environment, see [contributing.md](contributing.md)."

File: docs/source/user_guide/unit_testing.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+QD_WANTED_ARCHS=metal,vulkan python tests/run_tests.py
+
+# rerun the last failing tests first
+python tests/run_tests.py -f
+
+# stop at the first failure
+python tests/run_tests.py -x
+```
+
+The target architecture can also be set via `QD_WANTED_ARCHS` (comma-separated; supports `^arch` to exclude rather than include). Per-test timeouts default to 600 s and are enforced by `pytest_hardtle`, a CFFI-compiled C watchdog that can kill tests hung in native GPU calls even when the GIL is held.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: Mmove "Per-test timeouts default to 600 s and are enforced by `pytest_hardtle`, a CFFI-compiled C watchdog that can kill tests hung in native GPU calls even when the GIL is held." into an 'Advnaced' section

File: docs/source/user_guide/unit_testing.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+## Kernel compilation cache
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: move '## Kernel compilation cache' into an Advanced section.

File: docs/source/user_guide/unit_testing.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+During each test session the kernel compilation cache lives in a fresh, empty temp directory created by pytest's [`tmp_path_factory`](https://docs.pytest.org/en/stable/how-to/tmp_path.html) — typically `/tmp/pytest-of-<user>/pytest-<N>/qdcache0/`. Old session directories are cleaned up automatically by pytest's retention policy. This cache is separate from the user-facing `~/.cache/quadrants/` cache, and avoids recompiling identical kernels after each `qd.reset()` / `qd.init()` cycle within a session.
+
+## Per-file timing breakdown
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: move '## Per-file timing breakdown' into an 'Advanced' section

File: docs/source/user_guide/unit_testing.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+This is enabled by default in the Mac CI job; the results appear in the GitHub Actions job summary and are the primary tool for identifying slow test files.
+
+## Markers
+
+Tests can opt into two project-specific markers, in addition to pytest's built-in ones (`skip`, `xfail`, etc.).
+
+### `@pytest.mark.slow`
+
+Marks a test (or, more commonly, a specific `pytest.param(...)` case inside a parametrize list) as **slow** — long enough that the default test suite skips it. `tests/run_tests.py` adds `-m "not slow"` to the pytest invocation by default; pass `--run-slow` to opt back in:
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: - remove '(or, more commonly, a specific `pytest.param(...)` case inside a parametrize list)'

Comment (hughperkins) [reply docs/source/user_guide/unit_testing.md]: - remove '— long enough that the default test suite skips it'

File: docs/source/user_guide/unit_testing.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+# include slow
+python tests/run_tests.py --run-slow
+
+# slow ONLY (e.g. nightly job)
+python tests/run_tests.py -m slow --run-slow
+```
+
+The marker is used in two patterns:
+
+1. **Whole-test slow**: rare. The whole test always takes a long time and there's no smaller variant.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: - remove `rare`.
- remove `always`
- remove `and there's not smaller variant`

File: docs/source/user_guide/unit_testing.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+   ```python
+   @pytest.mark.slow
+   def test_thing_that_is_always_slow():
+       ...
+   ```
+
+2. **Slow-marked parametrize case** (preferred when applicable): a test parametrizes over a size axis and the large value is slow but the small value is cheap. The small value stays in the default suite as a smoke test; the large value moves to the slow lane. This is the dominant pattern in `tests/python/test_eig.py`, `test_linalg.py`, `test_ad_gdar_diffmpm.py`, etc.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: remove '(preferred when applicable)'

Comment (hughperkins) [reply docs/source/user_guide/unit_testing.md]: remove 'a test parametrizes over a size axis and the large value is slow but the small value is cheap. The small value stays in the default suite as a smoke test; the large value moves to the slow lane. This is the dominant pattern in `tests/python/test_eig.py`, `test_linalg.py`, `test_ad_gdar_diffmpm.py`, etc.'

File: docs/source/user_guide/unit_testing.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+   With this pattern the default suite still exercises the code path; the slow lane just adds the larger-size variant for full coverage.
+
+### `@pytest.mark.sample(...)`
+
+Marks a single heavily-parametrized test as opting in to **per-run stochastic sub-selection** of its parametrize cases. Use when:
+
+- the test's parametrize space is large (≥ ~16 cases),
+- each parametrize case is roughly independent (covering an independent corner case rather than a single bug class),
+- running every case every CI run is overkill, and
+- coverage convergence over many runs is acceptable for that test.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: - replace 'coverage convergence' with 'asymptotic coverage'
- remove 'for that test'

File: docs/source/user_guide/unit_testing.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+Apply it like any other marker, above the existing parametrize stack:
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: does it have to be above the stack? (if so it's not really 'like any other marker...)

File: docs/source/user_guide/unit_testing.md [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+@pytest.mark.sample(fraction=0.25)           # keep 25% of cases per run, min 1
+@pytest.mark.parametrize("size", [...])
+@pytest.mark.parametrize("dtype", [...])
+@pytest.mark.parametrize("layout", [...])
+@test_utils.test(arch=qd.gpu)
+def test_thing(size, dtype, layout):
+    ...
+```
+
+**Convergence math.** With `keep_n / total = k / N`, the probability that a *specific* parametrize case has been hit after `r` runs is `1 - (1 - k/N)^r`. For `n=6` out of 32 (`test_tile16_load_store`): ~65% after 5 runs, ~88% after 10, ~98% after 20. Combined with our CI cadence and the fact that any persistent bug surface lights up across multiple PRs, this gives effectively full coverage on a many-PR horizon at a fraction of the per-PR cost.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: remove this paragraph

File: docs/source/user_guide/unit_testing.md [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+   ```
+   python tests/run_tests.py --no-sample
+   ```
+
+**Per-test RNG independence.** Each `@sample`-marked test's subsample is seeded from `(global_seed, test_nodeid_prefix)`, so adding / renaming / tweaking the mark on `test_A` does NOT shift the sample of `test_B`. Routine refactors don't cause samples to migrate file-wide.
+
+**Composition with `slow`.** Sampling runs **after** marker-based filtering. With `--run-slow` not passed (the default), slow-marked parametrize cases drop out first, then the sampler sub-selects from the remaining (fast) cases. The intersection is the right composition: `--no-sample --run-slow` is the truly-exhaustive combo.
+
+**xdist note.** The seed is picked on the controller in `pytest_configure` (not in the per-worker `pytest_collection_modifyitems`), so all xdist workers see the same seed and produce the same sample. This is intentional — without this, each worker would subsample independently and `--sample-seed=<S>` wouldn't reproduce.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: move to 'advanced' section

File: docs/source/user_guide/unit_testing.md [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+
+**When *not* to use `@sample`.** If the test's parametrize axes are not roughly independent — e.g. axis A's bug surface only lights up when axis B is at a specific value — sampling can miss the interaction. Use `@slow` on the expensive subset instead, and keep the full Cartesian product for the cheap subset.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: remove this paragarph

File: docs/source/user_guide/unit_testing.md [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+```
+
+`@test_utils.test` is what wires the test into the per-backend matrix and applies platform exclusions (`exclude=`), extension requirements (`require=`, e.g. `qd.extension.data64` for f64 tests), and per-test options (`default_fp`, `fast_math`, etc.). See `tests/test_utils.py` for the full surface.
+
+Common helpers in `tests/test_utils.py`:
+
+- `test_utils.skip_if_f64_unsupported(dtype)` — skip the current test at runtime if `dtype == qd.f64` and the active backend can't carry f64 through buffer I/O (Metal, MoltenVK on Darwin). Use inside a parametrized test that sweeps both f32 and f64.
+- `test_utils.expected_archs()` — list of archs that the current `QD_WANTED_ARCHS` allows. Used to skip tests with no satisfiable arch.
+
+## CI checks
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: remove this section

File: docs/source/user_guide/contributing.md [hunk 16]
```diff
@@ -2,25 +2,13 @@
...
-```
-
-The target architecture can also be set via the `QD_WANTED_ARCHS` environment variable (comma-separated, e.g. `QD_WANTED_ARCHS=metal,vulkan`).
-
-### Kernel compilation cache
-
-During test runs, compiled kernels are cached to disk so that the same kernel is not recompiled after each `qd.reset()`/`qd.init()` cycle.
-
-A fresh, empty cache directory is created for each test session by pytest's [`tmp_path_factory`](https://docs.pytest.org/en/stable/how-to/tmp_path.html) (typically under `/tmp/pytest-of-<user>/pytest-<N>/qdcache0/`). Old session directories are cleaned up automatically by pytest's retention policy. This cache is separate from the user-facing `~/.cache/quadrants/` cache.
+See [unit_testing.md](unit_testing.md) for the full reference: launcher flags, env vars, the `slow` / `sample` markers, how to reproduce a single failing parametrize case, and how to do exhaustive runs for release gates.
```
Comment (hughperkins) [inline docs/source/user_guide/contributing.md]: - remove 'for the full reference: launcher flags, env vars, the `slow` / `sample` markers, how to reproduce a single failing parametrize case, and how to do exhaustive runs for release gates.'

File: docs/source/user_guide/unit_testing.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+
+2. **Slow-marked parametrize case**:
+
+   ```python
+   @pytest.mark.parametrize("n", [4, pytest.param(12, marks=pytest.mark.slow)])
+   def test_sym_eig_general(n):
+       ...
+   ```
+
+   With this pattern the default suite still exercises the code path; the slow lane just adds the larger-size variant for full coverage.
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: replace 'With this pattern' with 'In this specific example'

File: docs/source/user_guide/unit_testing.md [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+### `@pytest.mark.sample(...)`
+
+Marks a single heavily-parametrized test as opting in to **per-run stochastic sub-selection** of its parametrize cases. Use when:
+
+- the test's parametrize space is large (≥ ~16 cases),
+- each parametrize case is roughly independent (covering an independent corner case rather than a single bug class),
+- running every case every CI run is overkill, and
+- asymptotic coverage over many runs is acceptable.
+
+Apply it like any other marker. Position within the decorator stack is irrelevant — pytest attaches function-level markers to the test regardless of order, so `@pytest.mark.sample` can sit anywhere above or below the `@pytest.mark.parametrize` decorators:
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: remove ' Position within the decorator stack is irrelevant — pytest attaches function-level markers to the test regardless of order, so `@pytest.mark.sample` can sit anywhere above or below the `@pytest.mark.parametrize` decorators:'

File: docs/source/user_guide/unit_testing.md [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+@pytest.mark.sample(fraction=0.25)           # keep 25% of cases per run, min 1
+@pytest.mark.parametrize("size", [...])
+@pytest.mark.parametrize("dtype", [...])
+@pytest.mark.parametrize("layout", [...])
+@test_utils.test(arch=qd.gpu)
+def test_thing(size, dtype, layout):
+    ...
+```
+
+**How to reproduce.** Three levels of reproducibility:
```
Comment (hughperkins) [inline docs/source/user_guide/unit_testing.md]: 'How to reproduce' => 'How to reproduce failing tests'

File: docs/source/user_guide/unit_testing.md [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+   @pytest.mark.parametrize("n", [4, pytest.param(12, marks=pytest.mark.slow)])
+   def test_sym_eig_general(n):
+       ...
+   ```
+
+   In this specific example the default suite still exercises the code path; the slow lane just adds the larger-size variant for full coverage.
+
+### `@pytest.mark.sample(...)`
+
+Marks a single heavily-parametrized test as opting in to **per-run stochastic sub-selection** of its parametrize cases. Use when:
```
Comment (duburcqa) [inline docs/source/user_guide/unit_testing.md:81]: I don't like this idea at all. But you do what you want.

File: docs/source/user_guide/unit_testing.md [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+- the test's parametrize space is large (≥ ~16 cases),
+- each parametrize case is roughly independent (covering an independent corner case rather than a single bug class),
+- running every case every CI run is overkill, and
+- asymptotic coverage over many runs is acceptable.
```
Comment (duburcqa) [inline docs/source/user_guide/unit_testing.md:86]: I don't think it is ever the case in my opinion.

File: docs/source/user_guide/unit_testing.md [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+
+   In this specific example the default suite still exercises the code path; the slow lane just adds the larger-size variant for full coverage.
+
+### `@pytest.mark.sample(...)`
+
+Marks a single heavily-parametrized test as opting in to **per-run stochastic sub-selection** of its parametrize cases. Use when:
+
+- the test's parametrize space is large (≥ ~16 cases),
+- each parametrize case is roughly independent (covering an independent corner case rather than a single bug class),
+- running every case every CI run is overkill, and
```
Comment (duburcqa) [inline docs/source/user_guide/unit_testing.md:85]: I'm more in favour of nightly vs PR jobs. With a defined selection of required subset running on all PRs.

File: docs/source/user_guide/unit_testing.md (hunk 20 above)
Comment (hughperkins) [reply docs/source/user_guide/unit_testing.md:81]: Your concern is that when one makes a PR, and tests run multiple times, sometimes they might succeed, sometimes they might fail?

Thoughts on a compromise where the random seed is deterministically fixed by the name of the branch? So, re-running the tests on a specific PR won't change which tests run.

**Comment by hughperkins** [comment]
Mac tests are fast now => closing
